tanjun/repeaters.py

```diff
@@ -0,0 +1,425 @@
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""A way to add repeating tasks to Tanjun bots."""
+from __future__ import annotations
+
+__all__: list[str] = [
+    "AbstractRepeater",
+    "CallbackSig",
+    "CallbackSigT",
+    "Repeater",
+    "with_ignored_exceptions",
+    "with_fatal_exceptions",
+]
+
+import abc
+import asyncio
+import collections.abc
+import datetime
+import typing
+
+CallbackSig = collections.abc.Callable[..., collections.abc.Awaitable[None]]
+CallbackSigT = typing.TypeVar("CallbackSigT", bound=CallbackSig)
+
+
+class AbstractRepeater(abc.ABC):
+    """Abstract repeater class."""
+
+    __slots__ = ()
+
+    @property
+    @abc.abstractmethod
+    def iteration_count(self) -> int:
+        """Return the number of times this repeater has run.
+
+        Returns
+        -------
+        int
+            The number of times this repeater has run. Increments after
+            the callback is called, regardless if it was successful or not.
+        """
+
+    @abc.abstractmethod
+    def start(self) -> asyncio.Task[None]:
+        """
+        Start the repeater.
+
+        Returns
+        -------
+        asyncio.Task
+            The task running the actual repeating loop.
+        """
+
+    @abc.abstractmethod
+    def stop(self) -> None:
+        """Stop the repeater."""
+
+    @property
+    @abc.abstractmethod
+    def callback(self) -> CallbackSig:
+        """Return the callback attached to the repeater.
+
+        Returns
+        -------
+        CallbackSig
+            The callback attached to this repeater
+        """
+
+
+class Repeater(typing.Generic[CallbackSigT], AbstractRepeater):
```

Comment:
need a `as_repeater` decorator method which acts similarly to stuff `as_message_command`/`as_slash_command`